addons/portal/controllers/portal.py

```diff
@@ -228,6 +230,579 @@ def account(self, redirect=None, **post):
+        for key, new_val in address_values.items():
+            val = ResPartner._fields[key].convert_to_cache(partner[key], ResPartner)
+            if new_val != val and (val or new_val):
+                # Skip falsy values if unset in values and on record
+                return False
+        return True
+
+    def _handle_extra_form_data(self, extra_form_data, address_values):
+        """ Handling extra form data that were not processed on the address from.
+
+        :param dict extra_form_data: The extra form data.
+        :param dict address_values: The address value.
+        :return: None
+        """
+        pass
+
+    @route(
+        ['/portal/country_info/<model("res.country"):country>'],
+        type='jsonrpc',
+        auth="public",
+        methods=['POST'],
+        website=True,
+        readonly=True,
+    )
+    def shop_country_info(self, country, address_type, **kw):
+        address_fields = country.get_address_fields()
+        if address_type == 'billing':
+            required_fields = self._get_mandatory_billing_address_fields(country)
+        else:
+            required_fields = self._get_mandatory_delivery_address_fields(country)
+        return {
+            'fields': address_fields,
+            'zip_before_city': (
+                'zip' in address_fields
+                and address_fields.index('zip') < address_fields.index('city')
+            ),
+            'states': [(st.id, st.name, st.code) for st in country.sudo().state_ids],
+            'phone_code': country.phone_code,
+            'required_fields': list(required_fields),
+        }
+
+    @route('/address/set_as_default', type='jsonrpc', auth='user', website=True)
+    def address_set_as_default(self, address_id, address_type, **kw):
+        partner_sudo = self.env['res.partner'].sudo()
+        address_sudo = partner_sudo.browse(int(address_id))
+        # To-do: check with VFE if it is better to pass invoice as address_type from card instead
+        # each time check if billing then invoice means renaming of address_type in all the places
+        if address_type == 'billing':
+            address_type = 'invoice'
+        if not address_sudo._can_edited_by_current_customer():
+            raise Forbidden()
+        related_addresses = partner_sudo.search([
+            ('id', 'child_of', address_sudo.commercial_partner_id.ids),
+            '|',
+            ('type', 'in', [address_type, 'other']),
+            ('id', '=', address_sudo.commercial_partner_id.id),
+        ])
+        if address_type == 'invoice':
+            related_addresses.filtered(
+                'is_default_invoice_address'
+            ).is_default_invoice_address = False
+            address_sudo.is_default_invoice_address = True
+        else:
+            related_addresses.filtered(
+                'is_default_delivery_address'
+            ).is_default_delivery_address = False
+            address_sudo.is_default_delivery_address = True
```

Comment:
We shouldn't have to update n records everytime (and it won't happen if you switch to the logic we spoke about).